Add `Preface` chapter (#184)
esp-rs/book: src/preface.md

# Preface

> Welcome to a comprehensive guide to using the `esp-rs` driver, created for embedded Rust development on Espressif products. This book serves as a practical resource for developers wishing to utilise the capabilities of Rust in `no_std` environments when working with Espressif hardware platforms.

Author: Yeah, agree, sounds very confusing even for me now, but yesterday it felt ok-ish haha 😆

[modules]: https://docs.espressif.com/projects/rust/esp-hal/1.0.0-beta.0/esp32c6/esp_hal/index.html#modules

## Additional Resources

If you're unfamiliar with certain concepts covered in this book or would like to deepen your understanding, the following resources may be helpful:

| Resource                                               | Description                                                                          |
| ------------------------------------------------------ | ------------------------------------------------------------------------------------ |
| [The Rust Programming Language][rust-book]             | Learn Rust fundamentals before diving into embedded development.                     |
| [The Embedded Rust Book][embedded-rust-book]           | A collection of resources from Rust's Embedded Working Group.                        |
| [The Embedonomicon][embedonomicon]                     | Detailed insights into low-level embedded Rust programming.                          |
| [Embedded Rust (std) on Espressif][std-training]       | Getting started guide for using Rust's `std` environment on Espressif SoCs.         |

> | [Embedded Rust (no_std) on Espressif][no_std-training] | Guide for working in `no_std` environments with Espressif SoCs.                     |

Reviewer: Id recommend the no_std training before the std_training

Reviewer 2: Didn't we decide we weren't going to talk about `std` at all, since it's not supported by us and is probably just confusing?

<p style="text-align:center;"><img src="./assets/esp-rs.svg" width="25%"></p>

# Preface

> Welcome to a comprehensive guide to using the `esp-rs` ecosystem, created for embedded Rust development on Espressif products. This book serves as a practical resource for developers wishing to utilize the capabilities of Rust in `no_std` environments when working with Espressif hardware platforms.

Reviewer 2: Feels like we're repeating ourselves a bit here, we've already established in the first sentence that it's for use with Espressif products, I'm not sure we need to say this again in the very next sentence.

Author: It looked kind of "empty" with only one welcoming sentence, so I decided to add some quick overview of what will be covered in the book. Let me know if you want to change something there

> Welcome to a comprehensive guide to using the `esp-rs` ecosystem, created for embedded Rust development on Espressif products. This book serves as a practical resource for developers wishing to utilise the capabilities of Rust in `no_std` environments when working with Espressif hardware platforms.

Reviewer 2: ```suggestion
Welcome to a comprehensive guide to using the `esp-rs` ecosystem, created for embedded Rust development on Espressif products. This book serves as a practical resource for developers wishing to utilize the capabilities of Rust in `no_std` environments when working with Espressif hardware platforms.
```

## Who This Book Is For

> This book is intended for people with some experience in Rust and assumes basic knowledge of embedded development and electronics. The expected target user of our ecosystem and reader of this book has a basic understanding of concepts such as cross-compilation, common digital interfaces (UART, SPI, I2C), memory-mapped peripherals and interrupts. For those without prior experience, we recommend first reading the [Scope of This Book][prerequisites] and [Resources][resources] sections to get up to speed.

Reviewer 2: ```suggestion
This book is intended for people with some experience in Rust and assumes basic knowledge of embedded development and electronics. Readers of this book are expected to have a basic understanding of concepts such as cross-compilation, common digital interfaces (UART, SPI, I2C), memory-mapped peripherals and interrupts. For those without prior experience, we recommend first reading the [Scope of This Book][prerequisites] and [Resources][resources] sections to get up to speed.
```

This sounds a bit awkward, maybe something like I've suggested (or whatever else you come up with) would be a bit more natural?

> This book is intended for people with some experience in Rust and assumes a basic knowledge of embedded development and electronics. The expected target user of our product and reader of this book has a basic understanding of concepts such as cross-compilation, common digital interfaces (UART, SPI, I2C), memory-mapped peripherals and interrupts. For those without prior experience, we recommend first reading the [Scope of This Book][prerequisites] and [Resources][resources] sections to get up to speed.

Reviewer 2: ```suggestion
This book is intended for people with some experience in Rust and assumes basic knowledge of embedded development and electronics. The expected target user of our product and reader of this book has a basic understanding of concepts such as cross-compilation, common digital interfaces (UART, SPI, I2C), memory-mapped peripherals and interrupts. For those without prior experience, we recommend first reading the [Scope of This Book][prerequisites] and [Resources][resources] sections to get up to speed.
```

Also, what does "product" refer to here? I think this should be reworded as it's pretty ambiguous, IMO.

[prerequisites]: #scope-of-this-book
[resources]: #additional-resources

## Scope of This Book

In this book, we focus specifically on the `esp-rs` ecosystem. It does **not** serve as:

- A general Rust programming language tutorial: if you are new to Rust, we recommend starting with [The Rust Programming Language][rust-book].
- An embedded development guide: prior familiarity with basic embedded development is assumed.

> - A collection of examples: This book explains the overall process of working with the `esp-rs` ecosystem. More examples can be found [here][examples] or in the [Awesome ESP Rust][awesome-esp-rust] repository. 

Reviewer 2: Maybe rather than linking to the same resources in multiple places, we can just add a sentence after this list telling people to check the Additional Resources section below if they are not familiar with any of these concepts?

[examples]: https://github.com/esp-rs/esp-hal/tree/main/examples

## Stability and Availability

> The `esp-rs` project is under active development and may be subject to changes as it evolves. While we strive for stability, users should expect periodic modifications as we improve the API, improve performance, and introduce new features. Those [modules] that are already stabilised will not be subject to frequent dramatic changes, while the `unstable` parts of the drivers are being actively worked on and will likely undergo changes in future releases. 

Reviewer 2: ```suggestion
The `esp-rs` project is under active development and may be subject to changes as it evolves. While we strive for stability, users should expect periodic modifications as we improve the API, improve performance, and introduce new features. [Modules] that are already stabilized will not be subject to breaking changes, while the `unstable` parts of the drivers are being actively worked on and will likely undergo changes in future releases. 
```

> Additionally, the ecosystem of Rust itself, in a broader meaning, is still in its infancy. Some dependencies, tools, or language features may change over time. We encourage developers to stay up to date with the latest releases and participate in discussions to help shape the future of Rust on Espressif platforms.

Reviewer 2: ```suggestion
Additionally, the embedded Rust ecosystem is still in its infancy; some dependencies, tools, or language features may change over time. We encourage developers to stay up to date with the latest releases and participate in discussions to help shape the future of Rust on Espressif platforms.
```
